File: data/src/main/java/com/ykb/app/cryptotrader/data/model/TradeBot.java
package com.ykb.app.cryptotrader.data.model;

import com.ykb.app.cryptotrader.utils.enums.StrategyNames;
import com.ykb.app.cryptotrader.utils.enums.TradeBotStatus;
import jakarta.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.List;

@Entity
@Table(name = "TRADE_BOT")
@Data
@EqualsAndHashCode(callSuper = true)
public class TradeBot extends BaseIdEntity {

    @Column(name = "NAME", columnDefinition = "VARCHAR(256)", nullable = false)
    @Enumerated(EnumType.STRING)
    private StrategyNames strategyName;

    @Column(name = "STRATEGY_SETTINGS", columnDefinition = "VARCHAR(2048)", nullable = false)
    private String strategySettings;

    @ManyToOne
    @JoinColumn(name = "USER_ID", nullable = false)
    private User user;

    @Column(name = "STATUS", columnDefinition = "VARCHAR(16)")
    @Enumerated(EnumType.STRING)
    private TradeBotStatus status = TradeBotStatus.INIT;

    @OneToMany(mappedBy = "tradeBot", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TradeBotStateHistory> stateHistory;

    public TradeBot(User user, StrategyNames strategyName, String strategySettings) {
        this.user=user;
        this.strategyName=strategyName;
        this.strategySettings=strategySettings;
    }

    public boolean isInit() {
        return TradeBotStatus.INIT==status;
    }

    public boolean isRunning() {
        return TradeBotStatus.isRunning(status);
    }

    public boolean isTerminated() {
        return TradeBotStatus.TERMINATED==status;
    }

    public void terminate() {
        status = TradeBotStatus.TERMINATED;
        inactivate();
    }

    public void setStatus(TradeBotStatus status) {
        this.status = status;
        stateHistory.add(new TradeBotStateHistory(this, status));
    }

}

File: library/data/data-domain/src/main/kotlin/com/ykb/app/cryptotrader/data/domain/model/TradeBot.kt
package com.ykb.app.cryptotrader.data.domain.model

import com.ykb.app.cryptotrader.data.auth.model.User
import com.ykb.app.cryptotrader.data.base.entity.BaseIdEntity
import com.ykb.app.cryptotrader.utils.enums.TradeBotStatus
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table

@Entity
@Table(name = "TRADE_BOT")
class TradeBot(
    @ManyToOne
    @JoinColumn(name = "STRATEGY_ID", nullable = false)
    private val strategy: Strategy,

    @ManyToOne
    @JoinColumn(name = "USER_ID", nullable = false)
    private val user: User,

    @Column(name = "STATUS", columnDefinition = "VARCHAR(16)")
    @Enumerated(EnumType.STRING)
    private var status: TradeBotStatus = TradeBotStatus.ACTIVE
) : BaseIdEntity() {

    fun getStrategy(): Strategy {
        return strategy
    }

    fun getUser(): User {
        return user
    }

    fun isActive(): Boolean {
        return TradeBotStatus.ACTIVE==status
    }

    fun isTerminated(): Boolean {
        return TradeBotStatus.TERMINATED==status
    }

    fun terminate() {
        status = TradeBotStatus.TERMINATED
        inactivate()
    }

}